### DISCUSSION TITLE
Incorrect order of *Provider

### DISCUSSION THREAD

**Discussion by denissemyn** [resteasy/resteasy#2924] [Q&A] [answered]
Hi guys!
I use resteasy 2.3.6.Final 
And met strange issue with JSON parsing.
I send to server the JSON like:
{
    "prop1": "val1",
    "prop2":  "val2",
    "@class": "come.home.CommonDTO"
}
And have the response:
javax.xml.bind.UnmarshalException - with linked exception: [com.sun.istack.internal.SAXParseException2; columnNumber: 0; unexpected element (uri:"", local:"prop1"). Expected elements are <{}commonDTO>]

So seems that that Resteasy doesn't see JSON properties.
The most strange thing, that it works it my local PC. Bur is not working at remote tesing server.

I was debuged the code and find out, that for parsing JSON, resteasy uploads the list of MessageBodyReader, like:

class: ResteasyProviderFactory
method: getMessageBodyReader
List<ResteasyProviderFactory.SortedKey<MessageBodyReader>> readers = this.messageBodyReaders.getPossible(mediaType, type);

And I have different order of those MessageBodyReaders.

In my local PC is 

org.jboss.resteasy.plugins.providers.jackson.ResteasyJacksonProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JsonCollectionProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JsonMapProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonXmlRootElementProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonXmlSeeAlsoProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonXmlTypeProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonElementProvider
org.jboss.resteasy.plugins.providers.DataSourceProvider
org.jboss.resteasy.plugins.providers.StringTextStar
org.jboss.resteasy.plugins.providers.InputStreamProvider
org.jboss.resteasy.plugins.providers.ByteArrayProvider
org.jboss.resteasy.plugins.providers.FileProvider

But in remote server the order is different:

org.jboss.resteasy.plugins.providers.jaxb.json.JettisonXmlRootElementProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JsonCollectionProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JsonMapProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonXmlSeeAlsoProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonXmlTypeProvider
org.jboss.resteasy.plugins.providers.jaxb.json.JettisonElementProvider
org.jboss.resteasy.plugins.providers.jackson.ResteasyJacksonProvider
org.jboss.resteasy.plugins.providers.DataSourceProvider
org.jboss.resteasy.plugins.providers.StringTextStar
org.jboss.resteasy.plugins.providers.InputStreamProvider
org.jboss.resteasy.plugins.providers.ByteArrayProvider
org.jboss.resteasy.plugins.providers.FileProvider

Somehow this JettisonXmlRootElementProvider comes at fist position and breaks the JSON parsing.

I came deeper and foud that Providers are filling in org.jboss.resteasy.core.MediaTypeMap.

They are uploading and sorting:

Collections.sort(matches, new MediaTypeMap.TypedEntryComparator(type));
        cached = convert(matches);
        if (useCache) {
            this.classCache.put(cacheEntry, cached);
        } 
But! sorting works correctly, and seems like the registration of Providers is incorrect:
class: RegisterBuiltin       
method: registerProviders
Enumeration<URL> en = Thread.currentThread().getContextClassLoader().getResources("META-INF/services/" + Providers.class.getName());

And the order of this resources is different in different environments!
So that's the case.
But I have no idea why this order is different...

PS.
In all enviromnents openjdk version "1.8.0_265"

**Comment by jamezp** [comment, marked as answer]
Are you targeting a specific container? That is a quite old version of RESTEasy.

**Reply by denissemyn** [reply]
The application works inside Tomcat 8, and inside Docker container. It's quite old "legacy" application... 
Will migrate to newer version of Resteasy and see how it works.
Just curious what could cause different order of Providers registration.

**Reply by jamezp** [reply]
I don't know when it was introduced, but there is support for the `javax.annotation.Priority` annotation you can place on a provider. The Jettison providers were removed in RESTEasy 4.x. The 2.3.6 release is 9 years old now so an upgrade might be useful if it's possible for you :)

**Reply by denissemyn** [reply]
Updated rest-easy to 3.15.2.Final
Also added    export CATALINA_OPTS=-Djackson.deserialization.whitelist.packages=com.home.*
Now it works...
